pkg/scheduler/api/compatibility/compatibility_test.go

```diff
@@ -24,29 +24,77 @@ import (
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes/fake"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
 	_ "k8s.io/kubernetes/pkg/apis/core/install"
+	"k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/scheduler"
+	"k8s.io/kubernetes/pkg/scheduler/algorithmprovider"
 	_ "k8s.io/kubernetes/pkg/scheduler/algorithmprovider/defaults"
 	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
 	kubeschedulerconfig "k8s.io/kubernetes/pkg/scheduler/apis/config"
 	schedulerconfig "k8s.io/kubernetes/pkg/scheduler/apis/config"
 	"k8s.io/kubernetes/pkg/scheduler/core"
 )
 
-func TestCompatibility_v1_Scheduler(t *testing.T) {
-	snapshot := scheduler.RegisteredPredicatesAndPrioritiesSnapshot()
-	defer scheduler.ApplyPredicatesAndPriorities(snapshot)
+type testCase struct {
+	JSON             string
+	wantPredicates   sets.String
+	wantPrioritizers sets.String
+	wantPlugins      map[string][]kubeschedulerconfig.Plugin
+	wantExtenders    []schedulerapi.ExtenderConfig
+}
+
+func TestCompatibilityWithDefaultFeatureGates(t *testing.T) {
+	defer algorithmprovider.ApplyFeatureGates()()
+
+	testCompatibility(t, nil)
+}
+
+func TestCompatibilityWithAlphaFeature(t *testing.T) {
+	// Enable alpha feature EvenPodsSpread
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.EvenPodsSpread, true)()
+	defer algorithmprovider.ApplyFeatureGates()()
+
+	alphaFeatureTestCase := &testCase{
+		JSON: `{
+	  "kind": "Policy",
+	  "apiVersion": "v1",
+	  "predicates": [
+		{"name": "EvenPodsSpread"}
+	  ],
+	  "priorities": [
+		{"name": "EvenPodsSpreadPriority",   "weight": 2}
+	  ]
+	}`,
+		wantPrioritizers: sets.NewString(
+			"EvenPodsSpreadPriority",
+		),
+		wantPlugins: map[string][]kubeschedulerconfig.Plugin{
+			"FilterPlugin": {
+				{Name: "NodeUnschedulable"},
+				{Name: "TaintToleration"},
+				{Name: "PodTopologySpread"},
+			},
+		},
+	}
 
+	testCompatibility(t, alphaFeatureTestCase)
+}
+
+func TestCompatibilityWithBetaFeature(t *testing.T) {
+	// Disable beta feature AttachVolumeLimit
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.AttachVolumeLimit, false)()
+	defer algorithmprovider.ApplyFeatureGates()()
+
+	testCompatibility(t, nil)
+}
+
+func testCompatibility(t *testing.T, customTestCase *testCase) {
 	// Add serialized versions of scheduler config that exercise available options to ensure compatibility between releases
-	schedulerFiles := map[string]struct {
-		JSON             string
-		wantPredicates   sets.String
-		wantPrioritizers sets.String
-		wantPlugins      map[string][]kubeschedulerconfig.Plugin
-		wantExtenders    []schedulerapi.ExtenderConfig
-	}{
+	schedulerFiles := map[string]testCase{
```

Comment:
My idea is sth like this

```go
testCases := []struct {
	name             string
	JSON             string
	wantPredicates   sets.String
	wantPrioritizers sets.String
	wantPlugins      map[string][]kubeschedulerconfig.Plugin
	wantExtenders    []schedulerapi.ExtenderConfig
	setFeatureGates  func()
}{
	{
		name: "<describe purpose of this testcase>",
		......
		setFeatureGates: func() {
			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.EvenPodsSpread, true)()
		},
	}
}

for _, tc := range testCases {
	t.Run(tc.name, func(t *testing.T) {
		if tc.setFeatureGates != nil {
			tc.setFeatureGates()
		}
		defer algorithmprovider.ApplyFeatureGates()()
		......
	}
}
```

Reply:
Thanks for the detailed explanation, code is ready for review~

PS: If we call `defer featuregatetesting.SetFeatureGateDuringTest` in `setFeatureGates`, the feature gate will be restored when the func `setFeatureGates` ends, so we need to return this func in `SetFeatureGateDuringTest`. Now I'm using a map to set the feature gate in test case instead.

Reply:
Ah, defer has that trick.

Using map is fine. Or (not necessary in this case), to change the signature to `setFeatureGates  func() (restoreFunc func())`.